### ISSUE TITLE
Auction House item listing

### ISSUE THREAD

**Issue by Trisjdc** [TrinityCore/TrinityCore#15648]
Rev: https://github.com/TrinityCore/TrinityCore/commit/5fd42e68aeeee84bd04ad750a848460adead953d

When doing client-side sorting of the Auction House list results, sorting isn't working properly. It's only sorting the results for a given page.

A sort list is sent to the server on CMSG_AUCTION_LIST_ITEMS, it's this block of data:

https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Handlers/AuctionHouseHandler.cpp#L716

By looking at the received data, a list of <sortType, isDescending> is sent.

```
enum AuctionSortTypes
{
    AUCTION_SORT_BY_LEVEL       = 0,
    AUCTION_SORT_BY_RARITY      = 1,
    AUCTION_SORT_BY_TIME_LEFT   = 3,
    AUCTION_SORT_BY_SELLER      = 7,
    AUCTION_SORT_BY_CURR_BID    = 8,
};
```

A sort list has to be passed down to AuctionHouseObject::BuildListAuctionItems

**Comment by BlakeTNC** [comment]
![Bounty Amount Badge](https://api.bountysource.com/badge/issue?issue_id=27117179)
https://www.bountysource.com/issues/27117179-3-3-5-auction-house-item-listing

**Comment by BlakeTNC** [comment]
Hi All, 
I wondered if one of you developers has the required skill set and the interest to start the work on this issue? 
-Blake

Including a few of the other active devs: 
@Shauren,  @Vincent-Michael, @joschiwald, @Aokromes, @Gacko, @Rushor, @tkrokli

**Comment by tkrokli** [comment]
I don't know the solution here. Otherwise I would have posted a suggested solution already and more than a year ago.

**Comment by r00ty-tc** [comment]
There are ways to solve this. I've just never considered it that important. Did retail really sort the whole list? I'm fairly sure this was why we had such addons that did this for us.

**Comment by tkrokli** [comment]
Good point. IIRC I always used addons to get AH listings to get the sorting correct and I also think I had this issue on retail if I didn't use AH addons.

**Comment by Shauren** [comment]
Yes, retails does this and sorting by price is quite important

**Comment by BlakeTNC** [comment]
> There are ways to solve this. I've just never considered it that important. Did retail really sort the whole list? 

Yes, Retail sorts the whole list, and did the same in WOTLK. Any returned results are sorted according to the selected sort criteria across all pages. This is also the intuitive expected behavior. 

For some players who use the auction house frequently and enjoy buying and selling, this can actually be pretty important.

When buying: If you are trying to get the best deal on any particular item, you need to sort by the current bid price. If the results are not sorted across all pages, then the items at the top of the first page of results may not be the lowest prices available. The magnitude of this problem increases with the number of listed items, in other words it increases with the popularity of the server.

When selling: Let's say you are selling lots of Linen Cloth, and you want to set your asking price close to the lowest available price. You do a search for Linen Cloth to find out what other people are asking, but once again if you sort by bid price, the prices on the first page may not be accurate until this issue is fixed. 

When I first started playing Wow, I had as much fun trading for profit in the auction house as I did in the rest of the game. Building wealth was fun for some players and improved crafting item liquidity. Without proper sorting,  buying and selling competitively is more difficult.

This also affects how professional the private server appears. Any bug in something used as frequently as the auction house is highly visible all players, new and old.

Just my two cents. I do understand that for anyone who doesn't get a kick out of trading, this wouldn't matter much.

-Blake

**Comment by BlakeTNC** [comment]
Besides Current Bid, sorting by Rarity or by Level can also be important enough whenever needed.

**Comment by r00ty-tc** [comment]
I think it can be done, but I think the extra processing load for large auction houses makes merging this https://github.com/TrinityCore/TrinityCore/pull/19251 highly advisable first.

@Shauren were there any issues with that? I took a look and it seems to lock the listings from happening while other auction update actions are happening.

**Comment by BlakeTNC** [comment]
This may be obvious but I thought I'd mention that there are at least two options for implementing this. (Though I have no familiarity with Trinity code). I'm not sure which method would be optimal for processor load.

1) Let the database software do the work. Build queries that include "sort by" (ascending or descending based on the clients selection), and "limit" parameters (the limit parameter in mysql can specify the first and last record sent, with each range corresponding to one page of results). Send each page to the client as needed.

2) Do the work in code. Grab all applicable results from the database. Do the sorting in code, grab the appropriate range for the page being viewed, and send that over to the client.

Both above items assume that only one page at a time is sent to the client, as needed. But I don't know if that's the case with wow. If each search causes all pages to be sent at once, then no limit clauses would be needed regardless of where the work is done.

-Blake

**Comment by BlakeTNC** [comment]
If you have the database do the work, which (might) possibly be a cleaner way to go, then I wonder if having an index on the sorted columns could improve sort/query speed. Having the DB reside in memory if feasible could speed DB actions as well.

**Comment by r00ty-tc** [comment]
The choice is simple (and already made for us). All auctions are in memory already.

**Comment by BlakeTNC** [comment]
Thanks @r00ty-tc

**Comment by r00ty-tc** [comment]
Have a try of https://github.com/r00ty-tc/TrinityCore/commit/513a648b944d61b0690ac80a99822c519f2d3c5e

It's just a proof of concept, not all columns are included, I think in real terms we'll need to move things like item, item poster name and minlevel into AuctionEntry so as not to be looking them up in the sorting function.

Probably it makes sense to perform a whole sort once and cache it while the player looks through pages. But, I wonder if that's correct operation?

The cheaper way, but wouldn't be supporting the whole sortorder would be to create multimaps for the primary order types and just iterate those (moving the cost to the add/remove operations).

I think for certain the auction house thread PR should be reviewed/merged first though. Whatever is done here it puts too much pressure on the main thread.

**Comment by r00ty-tc** [comment]
I'm wondering if we shouldn't be doing all of the following

1: Add a default search multimap for auctions in the default frame lookup sort order (so they're pretty much instant).
2: Update/merge the AH listing thread change by xinef1
3: If there's a backlog of requests in the now threaded AH list process, then increase the time allowed between queries (throttle time).
4: All other searches either partial_sorted (of course last pages would take the most effort). OR we can perhaps cache a recent search by a player. That is sort the entire list once and store it. If the player keeps querying the same signature, then keep passing the next page of the cached list (of course with protection for auctions deleted in the meantime, probably by storing auction ID rather than Auction pointer).

Item 4 would actually solve another problem I've seen with auction item managers (tradeskillmaster and the like). When scanning for auctions, and scanning for cancels, auctions added during the scan can really mess things up. If the list is static during a constant scan, then it would resolve this problem. But that's another story.

Anyway, some thoughts from others over the best way forward would be appreciated.

**Comment by BlakeTNC** [comment]
The current client behavior might be worth mentioning: 
When you log in to Dalaran wow, the default sort order appears to be Rarity(decreasing). If you change the sort order, IE to Current Bid(increasing), then it is remembered for the rest of the play session until you log out, and is applied to all other searches that you make that session unless changed. This is true even if you walk away from the auctioneer and return.

**Comment by r00ty-tc** [comment]
That's right, I checked that out already. The default is actually 7 sort columns. Although selecting bid and buyout will make it up to 9.

Which for overall sorting doesn't add to cost, but with a preset map we can't support multiple sort columns.

For info the default sort mode is fully:
```
Level 1 - SortOrder: AUCTION_SORT_RARITY (Ascending)
Level 2 - SortOrder: AUCTION_SORT_MINLEVEL (Descending)
Level 3 - SortOrder: AUCTION_SORT_ITEM (Ascending)
Level 4 - SortOrder: AUCTION_SORT_MINBIDBUY (Ascending)
Level 5 - SortOrder: AUCTION_SORT_STACK (Descending)
Level 6 - SortOrder: AUCTION_SORT_BID (Ascending)
Level 7 - SortOrder: AUCTION_SORT_TIMELEFT (Ascending)
```

Note that MINBIDBUY, in standard interface AND Auctioneer, isn't actually mapped to any column. Whereas auctioneer actually uses another sortorder for buyout sorting not even used in standard interface.

At any rate, in truth the client expects sort order can be any combination of the available sort orders and that's where we hit the issue of time to process. Probably we need to decide on "some" sorting but not fully as it should be, but maintain performance, or full sorting and we need to work on performance with all I've said above.

**Comment by BlakeTNC** [comment]
Interesting. I'm surprised that sub-sorting is allowed. I'm not sure exactly how the user can even directly select any sort options beyond Level 1. (In the client, you can only click one field header at a time.)

> Probably we need to decide on "some" sorting but not fully as it should be, but maintain performance, or full sorting and we need to work on performance with all I've said above.

I think this is a human judgment call. Personally it's hard to imagine that level 2+ sorting is going to impact the user experience, either in the standard interface or in auctioneer. Note that if the auctioneer add-on was highly dependent on sorting specifics, then it already would not be working correctly in Trinity because of the lack of current proper sorting implementation. In other words, even if you choose not to fully implement level 2+ sorting, it does not seem that we could make the situation any -worse- for auctioneer add-on users.

So, if you choose to allow full sorting on all levels, then I imagine a performance hit will be unavoidable. Still, it would be a valid choice if retail accuracy is paramount. I don't think I have anything more to say about that option here.

If you choose to focus on proper level 1 sorting only, then here is one idea for implementing sub sorts in a "reasonable" way, that will still allow for maximizing processor performance efficiency.

You could still create one multi-map for each sort possibility. Level 1 would match one of the available fields, as expected. For level 2+ you could make the implemented sort options a constant (always the same), regardless of what Level 2+ options were requested, by adapting the default sort order that you included above (excluding only the level 1 field from the adapted list).

So any time that someone requested a sort with Level 1 Bid ascending, the full sort would always be the following, regardless of what the client requests for level 2+:

Level 1 - SortOrder: AUCTION_SORT_BID (Ascending)
Level 2 - SortOrder: AUCTION_SORT_RARITY (Ascending)
Level 3 - SortOrder: AUCTION_SORT_MINLEVEL (Descending)
Level 4 - SortOrder: AUCTION_SORT_ITEM (Ascending)
Level 5 - SortOrder: AUCTION_SORT_MINBIDBUY (Ascending)
Level 6 - SortOrder: AUCTION_SORT_STACK (Descending)
Level 7 - SortOrder: AUCTION_SORT_TIMELEFT (Ascending)

(To avoid creating a second multimap for bid, returned results for BID Descending could be the reverse of the above map.)

Similarly, if the client sends a request for a sort with Level 1 MINLEVEL, then the returned sort order would always be:
Level 1 - SortOrder: AUCTION_SORT_MINLEVEL (Descending)
Level 2 - SortOrder: AUCTION_SORT_RARITY (Ascending)
Level 3 - SortOrder: AUCTION_SORT_ITEM (Ascending)
Level 4 - SortOrder: AUCTION_SORT_MINBIDBUY (Ascending)
Level 5 - SortOrder: AUCTION_SORT_STACK (Descending)
Level 6 - SortOrder: AUCTION_SORT_BID (Ascending)
Level 7 - SortOrder: AUCTION_SORT_TIMELEFT (Ascending)

(Level 1 MINLEVEL ascending requests would be drawn from the same multi-map in reverse order.)

You see, the multi-map for each field could simply be an adaptation of the default sort order as you specified, with only the level 1 item replaced. This way you could still have a small enough number of possible sort orders to create one multimap for each possibility, if that was desired. From what you said before, it sounds like this option could be implemented using 9 or less individual multimap collections.

Nine multi-maps is still a significant number, but I don't think it's too crazy of a number for the potential performance  advantages. This is just another idea for consideration. Remember that I'm not familiar with the code base and of course you should do whatever feels best.

**Comment by r00ty-tc** [comment]
Interestingly, I'm now wondering if it actually honours them all.

For standard auction windows, there's only fixed options. So 5 combinations available only.

However, auctioneer works a bit different. It will let you "stack" sort orders. If you buyout price until it shows buyout price total with ascending sign showing, then click Item. It will send Item as level 1, and then buyout price as 2. But, that's not a supported standard frame option.

So, I do wonder if with the 3,3,5 AH it was possible to "properly" sort by item name, for example. It's always a 3rd or lower level sort item. If it was possible (man how do we prove that? AH sniffs?) then, they supported all the sort orders possible. If it wasn't sorted fully across pages, then they only offered the 5 fixed order lists... Which does beg the question as to why they send explicitly the order they want..

**Comment by BlakeTNC** [comment]
>Which does beg the question as to why they send explicitly the order they want..

Good point. Well, network protocols are harder to change than server client-side implementations. They may have been simply trying to make their network protocol highly flexible, although later they may have realized that much flexibility was not needed.

I was looking at how cpp implements comparison: http://en.cppreference.com/w/cpp/concept/Compare
I was wondering if there is some standard or nonstandard way to "daisy chain" comparisons together for sub sorting, so that the compare function implementations for individual fields would not have to be duplicated for each possible sort mode. In other words, if a compare function needed to return 0 (indicating equality), is there some way that the function could first check to see if another level of comparison exists? (If another level exists, the comparison could be passed down the chain. If another level does not exist, then zero would be returned.)

How to implement the comparison functions properly in CPP is a bit outside of my depth. My core language is Java and I haven't programmed in CPP in years.

**Comment by BlakeTNC** [comment]
Even if you can't daisy chain the compare function, you would still only need to write nine or less implementations of Compare, so long as you're not supporting level 2+ customization.

**Comment by BlakeTNC** [comment]
From your comments it sounds like you might be leaning away from honoring all possible subsort combinations?

**Comment by r00ty-tc** [comment]
If you look at the branch on my repo I linked earlier https://github.com/r00ty-tc/TrinityCore/commits/ah_misc you'll see I have the sub sorting done. std::sort (and partial_sort) takes a comparison function  as an argument. Take a look at that, and try on the supported columns (most except rarity minbidbuy (bid and buy separate work fine though) and whatever sort order 4 is.

Now that example is only proof of concept (it does work). Efficiency (esp regarding the lookups needed for some of the sorts) is a real problem, and either way imagine a server with 5k auctions per faction (not unusual) and say among the cities a total of say 50 people at the AH. They're searching away. Now it might not be so bad. Most are searching for specific items, and the sort is applied after any filter (of course). But some might have addons that scan the whole AH. If several are doing so at the same time (in a competitive market it's not unusual) then you'll be seeing load.

I think, it might be acceptable IF we have the threaded listings. But it would have to be optional in worldserver.conf.

**Comment by r00ty-tc** [comment]
>From your comments it sounds like you might be leaning away from honoring all possible subsort combinations?

Just trying to weigh up under real load conditions how this would perform.

**Comment by BlakeTNC** [comment]
Yes, since I've never used the auctioneer add-on (or other auction add-ons), I can't guess whether a restricted set of sort possibilities would or would not be sufficient for those softwares and those players. You'd be better at figuring that out.

I guess the only way to test the performance impact of new code is to try the new code in a beta environment. Personally I'd hope this sorting fix would be optional only -if- it was first proven to adversely impact overall performance. I mean, if big servers like Dalaran or others chose to turn off sorting-across-pages, then this work could feel semi-pointless from the player perspective, including mine.

I can't evaluate your code changes myself. I don't have a CPP trinity IDE environment set up yet, and my CPP language skills need a refresh too. I'd need time to get up to speed and acclimated if I was to become more deeply involved in trinity work in general. 

I'd agree that the tested and desired threading changes should be committed first. That's just good sense.

**Comment by BlakeTNC** [comment]
Hmmm, if you really wanted to cover all potential needs: 1) you could implement restricted sorting (where Level 1 is the only customizable sort level) as a default implementation with nearly no performance impact. AND 2) You could implement fully customizable level 2+ sorting as well, with all its performance impact, but that option can be disabled from the config file as desired.

**Comment by BlakeTNC** [comment]
I mean it would be more work on your side, but it would cover every potential need that's been discussed so far, while letting the administrator choose to maximize performance or features.

**Comment by r00ty-tc** [comment]
> I mean, if big servers like Dalaran or others chose to turn off sorting-across-pages, then this work could feel semi-pointless from the player perspective, including mine.

Well, if they couldn't switch it off, and it was creating world diffs of 500ms+, then it would be an entirely other kind of problem.

> Hmmm, if you really wanted to cover all potential needs: 1) you could implement restricted sorting (where Level 1 is the only customizable sort level) as a default implementation with nearly no performance impact. AND 2) You could implement fully customizable level 2+ sorting as well, with all its performance impact, but that option can be disabled from the config file as desired.

In most cases, not really. How that sort method works is, std::sort and partial perform a quicksort, and for each comparison operation they pass two items to be considered. In this case I pass in the AuctionEntry pointer. So I have two AuctionEntry items to look at, I also pass in the sort info collected from the search packet.

So, the for loop will loop through all the sort info, from top level down. If the item isn't the same at level 1, it will instantly return and not even look at the other levels. The other levels only make a showing if the first level is the same.

Now, the main slow down is the collecting of info for the item to get the name. It's a bit complicated.

So, on the auction entry, we have the Auction item ID. So we can get the auction item pointer from that ID. But, the Item class doesn't have all the info we need. Some is in the template. So, we need to fetch the pointer to that from the item.

But what about the name? Well you can pull that from the template too. But, wait. What if it's "of the whatever" kind, oh we need to find that info and append to the string from dbc data.. OK, what if the player is from another locale? Well OK we look up the correct locale string... A fair bit of extra processing needed.

So, I think probably a pointer to strings for owner name, item name and also store item quality would be good enough.. Except for the locale point.

Probably you could keep the en_us names in there and then pull from locale if player is on another locale. Or use server locale configuration to decide which string to store.

So, mainly now I have a solution that actually works. But, it's all about optimization, and the threading..

**Comment by BlakeTNC** [comment]
It sounds like you have a good plan in mind.

One note. Your original plan description specified: 1: Add a default search multimap for auctions in the default frame lookup sort order (so they're pretty much instant).

It seems like the idea of that step is to make common search modes essentially instant. So, I might also consider looking at a running server to see which searches are most common under real usage, and potentially adding default sorted multimaps for those search modes as well (Maybe for the top few encountered sort combinations). IE I imagine that  most players are going to be clicking on the Bid column very frequently in the standard interface.

That way you could prevent all whole bunch of partial sort operations from ever being needed.

**Comment by BlakeTNC** [comment]
You know, in addition to the needs of the players and the administrators, any good implementation has to be something that makes sense to the programmer. IE you, not me, lol.

**Comment by BlakeTNC** [comment]
I assume you are already familiar with the following, but just in case you were not:
Each multimap can have its own implementation of the compare function. In Java this ability is known as a "Comparator" and in CPP it appears like this is done with the "Compare" parameter available in the multimap constructor. This means that if you want, you can have more than one listing multimap, but have each sorted a different way. http://en.cppreference.com/w/cpp/container/multimap/multimap (The constructor signature that can take to compare parameter is at the top of the page.)

I only mention this because your previous comments only referenced having one sorted multimap used for optimizing common searches. I didn't know if you knew that multiple compare implementations can be created as desired.

**Comment by r00ty-tc** [comment]
For now I am going with real time sorting to see how it goes. There are some other ways, but they will require some overall design. So I think this is a good start.

Work has been moved to a TrinityCore repo branch https://github.com/TrinityCore/TrinityCore/commits/3.3.5-auctionhouse which incorporates the threaded listing from @xinef1 and this sorting process.

Still very WIP, but ongoing.
